compiler/src/main/java/info/teksol/mindcode/mindustry/optimisation/GlobalOptimizer.java

```diff
@@ -61,7 +62,15 @@ protected int countInstructions(Predicate<LogicInstruction> matcher) {
     }
     
     protected void replaceInstruction(LogicInstruction original, LogicInstruction replaced) {
-        int index = program.indexOf(original);
-        program.set(index, replaced);
+        for (int index = 0; index < program.size(); index++) {
+            if (program.get(index) == original) {
+                program.set(index, replaced);
+                return;
+            }
+        }

```

Comment:
I don't understand the difference between the two code blocks? I see that the original code was not checking the result of `indexOf`, which can return -1 (I think) for not found values.

At the same time, this is checking for object reference, not object equality. I think this is correct, but I wanted to flag this.

Reply:
The reason is that LogicInstruction is a value object -- it has `equals` and `hashCode` methods based on its values. This means that two instructions can be identical even when they're two distinct instructions at different places in the program (eg. every `end` instruction is equal to every other). The previous version of the code therefore could end up replacing wrong instance of the instruction in the program, which resulted in one pretty mind-boggling bug :)

I'll add a comment to the method to explain the issue.

Btw, the DeadCodeEliminator also relies on selecting instructions for removal by their equality. It's not a problem there, because if an instruction is eligible for removal, so are all other copies of it. The subtlety was lost on me at first, though.